test/database_service/test_db_experiment_data.py

```diff
@@ -0,0 +1,724 @@
+
+        a_job = mock.create_autospec(Job, instance=True)
+        a_job.result.return_value = self._get_job_result(2)
+
+        called_back = False
+        callback_kwargs = "foo"
+        exp_data = DbExperimentData(backend=self.backend, experiment_type="qiskit_test")
+        exp_data.add_data(a_job, _callback, foo=callback_kwargs)
+        exp_data.block_for_results()
+        self.assertTrue(called_back)
+
+    def test_add_data_pending_post_processing(self):
+        """Test add job data while post processing is still running."""
+
+        def _callback(_exp_data, **kwargs):
+            kwargs["event"].wait(timeout=3)
+
+        a_job = mock.create_autospec(Job, instance=True)
+        a_job.result.return_value = self._get_job_result(2)
+
+        event = threading.Event()
+        self.addCleanup(event.set)
+
+        exp_data = DbExperimentData(experiment_type="qiskit_test")
+        exp_data.add_data(a_job, _callback, event=event)
+        with self.assertLogs("qiskit_experiments", "WARNING"):
+            exp_data.add_data({"foo": "bar"})
+
+    def test_get_data(self):
+        """Test getting data."""
+        data1 = []
+        for _ in range(5):
+            data1.append({"counts": {"00": randrange(1024)}})
+        results = self._get_job_result(3)
+
+        exp_data = DbExperimentData(experiment_type="qiskit_test")
+        exp_data.add_data(data1)
+        exp_data.add_data(results)
+        self.assertEqual(data1[1], exp_data.data(1))
+        self.assertEqual(data1[2:4], exp_data.data(slice(2, 4)))
+        self.assertEqual(
+            results.get_counts(), [sdata["counts"] for sdata in exp_data.data(results.job_id)]
+        )
+
+    def test_add_figure(self):
+        """Test adding a new figure."""
+        hello_bytes = str.encode("hello world")
+        file_name = uuid.uuid4().hex
+        self.addCleanup(os.remove, file_name)
+        with open(file_name, "wb") as file:
+            file.write(hello_bytes)
+
+        sub_tests = [
+            ("file name", file_name, None),
+            ("file bytes", hello_bytes, None),
+            ("new name", hello_bytes, "hello_again.svg"),
+        ]
+
+        for name, figure, figure_name in sub_tests:
+            with self.subTest(name=name):
+                exp_data = DbExperimentData(backend=self.backend, experiment_type="qiskit_test")
+                fn = exp_data.add_figures(figure, figure_name)
+                self.assertEqual(hello_bytes, exp_data.figure(fn))
+
+    @skipIf(not HAS_MATPLOTLIB, "matplotlib not available.")
+    def test_add_figure_plot(self):
+        """Test adding a matplotlib figure."""
+        # pylint: disable=import-error
+        import matplotlib.pyplot as plt
+
+        figure, ax = plt.subplots()
+        ax.plot([1, 2, 3])
+
+        service = self._set_mock_service()
+        exp_data = DbExperimentData(backend=self.backend, experiment_type="qiskit_test")
+        exp_data.add_figures(figure, save_figure=True)
```

Comment:
Does this actually write an svg file to disk? If so it'd be good to have a self.addCleanup call to remove the generated file if any (for local testing anyway).

Reply:
It doesn't. It uses a `mock` which does nothing other than letting me verify the proper methods are called.